pkg/controller/datavolume-controller.go

```diff
@@ -515,6 +697,278 @@ func (r *DatavolumeReconciler) sourceInUse(dv *cdiv1.DataVolume) (bool, error) {
+			nn := types.NamespacedName{Namespace: namespaces[i], Name: names[i]}
+			if err := r.client.Get(context.TODO(), nn, pvc); err != nil {
+				if !k8serrors.IsNotFound(err) {
+					return err
+				}
+			} else {
+				pod := &corev1.Pod{}
+				nn := types.NamespacedName{Namespace: namespaces[i], Name: expansionPodName(pvc)}
+				if err := r.client.Get(context.TODO(), nn, pod); err != nil {
+					if !k8serrors.IsNotFound(err) {
+						return err
+					}
+				} else {
+					if err := r.client.Delete(context.TODO(), pod); err != nil {
+						if !k8serrors.IsNotFound(err) {
+							return err
+						}
+					}
+				}
+
+				if err := r.client.Delete(context.TODO(), pvc); err != nil {
+					if !k8serrors.IsNotFound(err) {
+						return err
+					}
+				}
+			}
+		}
+	}
+
+	ot := &cdiv1.ObjectTransfer{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: name}, ot); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+	} else {
+		if err := r.client.Delete(context.TODO(), ot); err != nil {
+			if !k8serrors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+
+	RemoveFinalizer(dv, crossNamespaceFinalizer)
+	if err := r.client.Update(context.TODO(), dv); dv != nil {
+		return err
+	}
+
+	return nil
+}
+
+func expansionPodName(pvc *corev1.PersistentVolumeClaim) string {
+	return "cdi-expand-" + string(pvc.UID)
+}
+
+func (r *DatavolumeReconciler) expand(log logr.Logger, dv *cdiv1.DataVolume, pvc *corev1.PersistentVolumeClaim) (bool, error) {
+	requestedSize, hasRequested := dv.Spec.PVC.Resources.Requests[corev1.ResourceStorage]
+	currentSize, hasCurrent := pvc.Spec.Resources.Requests[corev1.ResourceStorage]
+	actualSize, hasActual := pvc.Status.Capacity[corev1.ResourceStorage]
+	if !hasRequested || !hasCurrent || !hasActual {
+		return false, fmt.Errorf("sizes missing")
+	}
+
+	expansionRequired := actualSize.Cmp(requestedSize) < 0
+
+	log.V(3).Info("Expand sizes", "req", requestedSize, "cur", currentSize, "act", actualSize, "exp", expansionRequired)
+
+	if expansionRequired && requestedSize.Cmp(currentSize) != 0 {
+		pvc.Spec.Resources.Requests[corev1.ResourceStorage] = requestedSize
+		if err := r.client.Update(context.TODO(), pvc); err != nil {
+			return false, err
+		}
+
+		return false, nil
+	}
+
+	podName := expansionPodName(pvc)
+	podExists := true
+	pod := &corev1.Pod{}
+	nn := types.NamespacedName{Namespace: pvc.Namespace, Name: podName}
+	if err := r.client.Get(context.TODO(), nn, pod); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return false, err
+		}
+
+		podExists = false
+	}
+
+	if !expansionRequired && !podExists {
+		// finally done
+		return true, nil
+	}
+
+	hasPendingResizeCondition := false
+	for _, cond := range pvc.Status.Conditions {
+		if cond.Type == corev1.PersistentVolumeClaimFileSystemResizePending {
+			hasPendingResizeCondition = true
+			break
+		}
+	}
+
+	if !podExists && !hasPendingResizeCondition {
+		// wait for resize condition
+		return false, nil
+	}
+
+	if !podExists {
```

Comment:
Can we get this part into its own function something like getOrCreateExpansionPod, the expand function is currently huge and hard to read.

Reply:
done